Add delayed datacheck to the datafeed job runner

benwtrent opened the pull request:
This adds a latent data check to the `DatafeedJob`when it runs real-time jobs. 

It can optionally be enabled via the `DatafeedConfig`.

Two new options added to the `DatafeedConfig` object, client side, and docs have been updated accordingly. 

Couple of decision points I could not really settle on:

* Should we set a default `delayed_data_check_window` if `should_run_delayed_data_check` is `true`?
* Should we put sane limits on `delayed_data_check_window` outside of the current bucket length checks? If so, what are those sane limits?
* Should we attempt to put the time of the missing data in the Audit? Maybe put the timestamp of the latest bucket missing data?
* How often should we check for this missing data?

The integration tests are done within `DelayedDataDetectorIT` previously included, that covered the actual looking up of missing data and doing real queries, etc. against a cluster.

Consequently, I just mocked out `DelayedDataDetector` in the `DatafeedJobTests` class and thought that coverage adequate. 

Screenshot of how it looks:

<img width="598" alt="screen shot 2018-11-09 at 9 00 22 am" src="https://user-images.githubusercontent.com/4357155/48269763-f87d6a80-e3fd-11e8-856c-6e0a589f4118.png">

Relates to https://github.com/elastic/elasticsearch/issues/35131

elasticmachine commented:
Pinging @elastic/ml-core

davidkyle commented on client/rest-high-level/src/test/java/org/elasticsearch/client/ml/datafeed/DatafeedConfigTests.java:
```diff
@@ -103,6 +103,13 @@ public static DatafeedConfig createRandom() {
         if (randomBoolean()) {
             builder.setChunkingConfig(ChunkingConfigTests.createRandomizedChunk());
         }
+        if (randomBoolean()) {
```
nit: My understanding is that you can have a time value (`setDelayedDataCheckWindow` ) set even if `setShouldRunDelayedDataCheck == false` because the check can be toggled on or off and that's the reason for using 2 fields rather than one. If that is the case the test should reflect that and remove the condition `if (shouldRunDelayedDataCheck) { // set the time window...`

davidkyle replied:
Same with DatafeedUpdateTests

benwtrent replied:
you are correct. I shall update the tests

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:
```diff
@@ -41,6 +43,7 @@
 
     private static final Logger LOGGER = LogManager.getLogger(DatafeedJob.class);
     private static final int NEXT_TASK_DELAY_MS = 100;
+    private static final int BUCKETS_BETWEEN_MISSING_DATA_CHECK = 1;
```
Too small? It could be a function of time rather than bucket count i.e check every 10 mins

benwtrent replied:
@davidkyle yeah, @dimitris-athanasiou originally brought up the idea of every bucket. 

How long between checks is still up for debate as the true impact on performance is not known.

davidkyle commented:
> Should we set a default delayed_data_check_window if should_run_delayed_data_check is true?

I say no. If we are using the model where the user has to set both params then it's a problem is both aren't set, otherwise the behaviour may be unexpected. I do think `should_run_delayed_data_check`  should be true be default with a sensible value of `delayed_data_check_window`

> Should we put sane limits on delayed_data_check_window outside of the current bucket length checks? If so, what are those sane limits?

With a 10 minute bucket the current value of 10,000 buckets is over 64 days. Did you consider making the limit an absolute time value rather than a function of the bucket length.

> Should we attempt to put the time of the missing data in the Audit? Maybe put the timestamp of the latest bucket missing data? 

Yes please. 
Is there anything that stops the same message being audited again and again each time the check is made?

> How often should we check for this missing data?

Less frequently 🤷‍♂️

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:
```diff
@@ -151,9 +158,29 @@ long runRealtime() throws Exception {
+    }
+
+    private boolean checkForMissingDataTriggered() {
+        return this.latestFinalBucketEndTimeMs > this.lastBucketDataCheckTimeMs * BUCKETS_BETWEEN_MISSING_DATA_CHECK;
```
Should this be 
`this.latestFinalBucketEndTimeMs > this.lastBucketDataCheckTimeMs + (bucketSpan * BUCKETS_BETWEEN_MISSING_DATA_CHECK);

benwtrent replied:
lulz, you are right, bad at math :D

benwtrent commented:
@davidkyle when it comes to a sane upper limit I can see benefits to either using a true TimeValue or a multiple of the bucket span.

I do realize that it is way too large right now, the current limit is dictated by the number of documents that are searchable at a time and not by a realistic use case. 

>Yes please.
Is there anything that stops the same message being audited again and again each time the check is made?

There is nothing preventing this right now...but if we store the latest timestamp of the bucket with missing docs && how many docs are missing overall, we could simply check if the values have changed and skip writing the message so that the audit log does not get spammed. 

> I do think `should_run_delayed_data_check` should be true by default with a sensible value of `delayed_data_check_window`

Any suggestions on a sensible value?

benwtrent commented:
@davidkyle Updated. Key updates

* Interval check is now 15 minutes, or the data check window length, which ever is smaller. If the querydelay+frequency is longer than both, we will be doing a check every time.
* Default values are 2 hours for window length, and true
* Fixed some of the initialization of the `DelayedDataChecker`, I did not like how I was validating the window, so adjusted it so that it is never null, and is initialized in the constructor. This leads to clearer behavior. 
* Added the timestamp to the audit message and prevented audit spamming.

benwtrent commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/datafeed/DatafeedConfig.java:
```diff
@@ -108,9 +113,16 @@ private static BytesReference parseBytes(XContentParser parser) throws IOExcepti
+     * The window of time to check for missing data
+     */
+    private final TimeValue delayedDataCheckWindow;
+    private final boolean shouldRunDelayedDataCheck;
```
This should be `Boolean` to allow null values so that the server side can default.

benwtrent commented:
Jenkins retest this please

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java:
```diff
@@ -151,9 +160,49 @@ long runRealtime() throws Exception {
+
+            // Keep track of the last bucket time for which we did a missing data check
+            this.lastDataCheckTimeMs = this.currentTimeSupplier.get();
+            List<DelayedDataDetector.BucketWithMissingData> response = delayedDataDetector.detectMissingData(latestFinalBucketEndTimeMs);
```
rename `response` to `missingDataBucket`?

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DelayedDataDetector.java:
```diff
@@ -38,24 +41,23 @@
-        if (Intervals.alignToFloor(windowMillis/bucketSpan, bucketSpan) >= 10000) {
-            throw new IllegalArgumentException("[window] must contain less than 10000 buckets at the current [bucket_span]");
+
+        if (datafeedConfig.getShouldRunDelayedDataCheck()) {
```
What if we didn't check this in the detector at all and we simply check this in the `DatafeedJob.checkForMissingDataIfNecessary()`? I think it would make this class cleaner. Alternatively, we could also have a `NullDelayedDataDetector` implementation that always returns an empty list when check is invoked. The latter would allow us to sanity check the `window` value too.

benwtrent commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ml/datafeed/DatafeedConfig.java:
```diff
@@ -62,6 +62,8 @@
     public static final ParseField AGGREGATIONS = new ParseField("aggregations");
     public static final ParseField SCRIPT_FIELDS = new ParseField("script_fields");
     public static final ParseField CHUNKING_CONFIG = new ParseField("chunking_config");
+    public static final ParseField DELAYED_DATA_CHECK_WINDOW = new ParseField("delayed_data_check_window");
```
Sounds good to me

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java:
```diff
@@ -519,6 +570,17 @@ public void setScrollSize(int scrollSize) {
         }
 
+        public void setDelayedDataCheckWindow(TimeValue delayedDataCheckWindow) {
+            if (delayedDataCheckWindow != null) {
```
Why should that matter? When doing the check, I round down the start/end values so that they themselves are multiples of the bucket_span.

dimitris-athanasiou replied:
in other similar cases we validate the window is a multiple of the bucket span to avoid the implicit handling of this. I don't think it's a problem here, I was asking to understand the behaviour.

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DelayedDataDetector.java:
```diff
@@ -38,24 +41,23 @@
+        if (datafeedConfig.getShouldRunDelayedDataCheck()) {
+            this.window = calculateWindowSize(datafeedConfig.getDelayedDataCheckWindow(), job.getAnalysisConfig().getBucketSpan());
+        } else {
+            this.window = 0; //implies that start == end and no queries are made
```
technically, yes, the comment is still accurate. But I see what you mean with a NullDelayedDataChecker, etc. to clean up some of this logic.

benwtrent commented:
Jenkins retest this please

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedJobValidator.java:
```diff
@@ -34,10 +34,15 @@ public static void validate(DatafeedConfig datafeedConfig, Job job) {
+
+    private static void checkValidDelayedDataCheckConfig(TimeValue bucketSpan, DelayedDataCheckConfig delayedDataCheckConfig) {
+        TimeValue delayedDataCheckWindow =  delayedDataCheckConfig.getCheckWindow();
+        if (delayedDataCheckWindow.equals(DelayedDataCheckConfig.DEFAULT_DELAYED_DATA_WINDOW) == false) {
```
Not sure it's worth splitting path here. The default value should also be a valid one so why not ignore it?

benwtrent replied:
What if the `bucket_span` is larger than `DelayedDataCheckConfig.DEFAULT_DELAYED_DATA_WINDOW` ?

As for the `window` spanning too many buckets, that is not possible with a 2 hour default and requiring `bucket_span` to be measured in seconds.

dimitris-athanasiou replied:
Hm, good point. But then we get in a situation where we cannot distinguish really between a default config and a user one set to 2 hours. Let's discuss this futher offline.

benwtrent replied:
We discussed this offline. 

The idea here is inherently broken. We should instead allow a `null` value for the window and then calculate the appropriate default based off of the `bucket_span` and other criteria. 

Working on adjusting this now.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/delayeddatacheck/DelayedDataDetectorFactory.java:
```diff
@@ -0,0 +1,107 @@
+
+public class DelayedDataDetectorFactory {
+
+    private static final int FALLBACK_NUMBER_OF_BUCKETS_TO_SPAN = 7;
```
Could you add a comment about why we selected 7 here?

benwtrent replied:
Because it is a luck number :D. I am going to rethink the default number of buckets and provide a better explanation for the default (hopefully soon).

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java:
```diff
@@ -177,8 +171,7 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalVInt(scrollSize);
         out.writeOptionalWriteable(chunkingConfig);
         if (out.getVersion().onOrAfter(Version.CURRENT)) {
```
That breaks BWC tests from what I can tell. I will change after backport.

benwtrent commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java:
```diff
@@ -212,11 +203,9 @@ public DatafeedConfig(StreamInput in) throws IOException {
         } else {
-            delayedDataCheckWindow = null;
-            shouldRunDelayedDataCheck = false;
+            delayedDataCheckConfig = DelayedDataCheckConfig.disabledDelayedDataCheckConfig();
```
I honestly don't know. The default makes sense so that we are consistent with initial creation. I do not fully grok the implications in a mixed versioned cluster however.

dimitris-athanasiou replied:
That makes 2 of us :-) I'd say let's go for the default for the reason you mention.

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/delayeddatacheck/DelayedDataDetectorFactory.java:
```diff
@@ -37,19 +38,21 @@ public static DelayedDataDetector buildDetector(Job job, DatafeedConfig datafeed
             }
+        }
+        if (currentWindow.compareTo(bucketSpan) < 0) {
+            // If it is the default value, we assume the user did not set it and the bucket span is very large
```
This could no longer be the default value as `currentWindow` is not null, right?

benwtrent replied:
HAHA! Hey man! Everybody knows you shouldn't read the comments :D 

I will remove it :)

benwtrent commented:
Jenkins retest this please
